Adopt new electricity tariffs in pvpc hourly pricing

azogue opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
With the change to the new, and unique, electric tariff 2.0TD, if you previously had configured multiple PVPC sensors monitoring prices for more than one of the old tariffs, only the first one will survive, so if you have any automations or scripts that depend on these removed sensors, you might need to adjust them.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

**Full adoption** of new electricity tariffs in Spain, operating since 2021-06-01, following #51320, which was a quick fix to adopt the new data schema, but had no new parameters expected by users, which the [new version](https://github.com/azogue/aiopvpc/releases/tag/v2.2.0) now implements 🎉

##### Details about the version bump:

Release notes: https://github.com/azogue/aiopvpc/releases/tag/v2.2.0
Changelog: https://github.com/azogue/aiopvpc/blob/master/CHANGELOG.md
Diff/compare against current version: https://github.com/azogue/aiopvpc/compare/v2.1.2...v2.2.0

##### Changes here

In this PR, in addition to the version bump for the external library, there are:

* **Changes on the current config flow** to reflect the new tariffs, and new parameters used by the sensor
* **Config entry migration** for painless jump to new parameters
* New **Options Flow** to edit those parameters
* And the necessary changes in tests to reflect all of that, including a new fixture with the new data schema

Also, the `CONFIG_SCHEMA` and the `async_setup` for the already deprecated YAML config (in 2021.6.0), **are not used anymore** in tests, so they're ready for removal :)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #51472
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/18179

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

azogue committed:
Bump aiopvpc from 2.1.1 to 2.2.0

azogue committed:
Update strings to match new UI schema and options flow

azogue committed:
Update translations generated with `script.translations develop`

azogue committed:
tests: add fixture with new data schema for new tariffs

azogue committed:
Use async_on_remove instead of async_will_remove_from_hass

azogue committed:
Update UI schema and add config entry migration

New tariffs are totally different to the old ones,
so in the migration logic the new attributes are filled by default,
and the user is notified with a warning, pointing to the options flow
to customize the new attributes.
(should be better to create a persist_notification?)

For users having more than one sensor!! (not really expected,
max 3 for the 3 old tariffs), only the first one survives,
and the others are completely removed (this behavior is reasonable, given
that all sensors are showing the same right now)

Also, 2 new attributes are included, to define the 2 contracted powers
(for different periods), which are now used in the sensor attributes

azogue committed:
Add options-flow to configure sensor parameters

azogue committed:
Update tests to check migration, new options-flow, and new data schema

* Use MockConfigEntry for setup, avoiding already deprecated YAML setup
* Check config migration happens as expected
* Adds a sensor parameters change with options flow,
  checking that tariff swap works ok

MartinHjelmare commented on homeassistant/components/pvpc_hourly_pricing/__init__.py:
```diff
@@ -52,3 +73,54 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+            ATTR_POWER: DEFAULT_POWER_KW,
+            ATTR_POWER_P3: DEFAULT_POWER_KW,
+        }
+        config_entry.version = 2
```
We try to avoid migrating the config entry version since that makes it impossible to role back for users. Can we detect the need to migrate by looking at the config entry data and just migrate the data as needed without changing config entry version?

azogue replied:
yes, of course, I'll change that 👍

azogue replied:
should I look at the `config_entry.options`, which will be empty for old entries, and fill it with the changes without increasing version?

azogue replied:
@MartinHjelmare, 

if  `VERSION = 1` remains in the ConfigFlow, then the `async_migrate_entry` is not called... did you suggest that I should avoid entirely the 'migrate-entry' flow and **just edit the entries inside the `async_setup_entry`** (in init or sensor)? I'm not sure if I understood correctly, sorry 😓

About this migration thing, I would also want to add a new unit test for that (coverage control is complaining about), but I'm having issues with the flow and the MockConfigEntry/MockEntity/MockEntityPlatform objects 😖. I know for sure that works in practice, but I think it should be reflected on tests too :)

MartinHjelmare replied:
Yes, in `async_setup_entry` in the init module, ie here in this module, we can just use `hass.config_entries.async_update_entry` to update the entry as we need to.

Yes, migrations are very good to test. Let me know if I can give some input on something for that. The standard way is to add a MockConfigEntry with the old data, and test that it's updated to the correct data state after the integration is setup. Same for the entity registry migration, set up the integration from old data and let the migration happen and then assert the new entity registry state.

azogue replied:
I moved the migration logic inside the setup, and works as expected for _real_ sensors: first one gets migration and others get removed, but I cannot reproduce that logic with the MockConfigEntry, as the second call to `await async_migrate_entries(hass, entry.entry_id, update_unique_id)` does not fail with ValueError, and instead I get the "Platform XXX does not generate unique IDs" error :( 

I suppose it's because the test doesn't handle a proper entity registry with old sensors, but I'm not sure what I'm doing wrong 😅

The test I have is as follows:
```python
async def test_multi_sensor_migration(
    hass, caplog, legacy_patchable_time, pvpc_aioclient_mock: AiohttpClientMocker
):
    """Test tariff migration when there are >1 old sensors."""
    hass.config.time_zone = dt_util.get_time_zone("Europe/Madrid")
    uid_1 = "discrimination"
    uid_2 = "normal"
    old_conf_1 = {CONF_NAME: "test_pvpc_1", ATTR_TARIFF: uid_1}
    old_conf_2 = {CONF_NAME: "test_pvpc_2", ATTR_TARIFF: uid_2}

    config_entry_1 = MockConfigEntry(domain=DOMAIN, data=old_conf_1, unique_id=uid_1)
    config_entry_1.add_to_hass(hass)
    config_entry_2 = MockConfigEntry(domain=DOMAIN, data=old_conf_2, unique_id=uid_2)
    config_entry_2.add_to_hass(hass)

    mock_data = {"return_time": datetime(2019, 10, 27, 20, tzinfo=date_util.UTC)}

    def mock_now():
        return mock_data["return_time"]

    with patch("homeassistant.util.dt.utcnow", new=mock_now):
        await hass.config_entries.async_setup(config_entry_1.entry_id)

       # here it raises an OperationNotAllowed, but the 'unique IDs' error is already fired ¿?
       with pytest.raises(Exception):
            await hass.config_entries.async_setup(config_entry_2.entry_id)

        # await hass.config_entries.async_setup(config_entry_2.entry_id)
        assert_setup_component(1, DOMAIN)

        # check migration
        current_entries = hass.config_entries.async_entries(DOMAIN)
        # assert len(current_entries) == 2
        for migrated_entry in current_entries:
            assert migrated_entry.version == 1
            assert migrated_entry.data[ATTR_POWER] == migrated_entry.data[ATTR_POWER_P3]
            assert migrated_entry.data[ATTR_TARIFF] == TARIFFS[0]

        await hass.async_block_till_done()
```

Tomorrow I'll try to find an example to follow to define an entity registry to reflect a real migration + deletion for the scenario with >1 sensors

azogue replied:
> ~~Tomorrow I'll try to find an example to follow to define an entity registry to reflect a real migration + deletion for the scenario with >1 sensors~~

I think I got it 👍

MartinHjelmare commented on homeassistant/components/pvpc_hourly_pricing/__init__.py:
```diff
@@ -52,3 +73,54 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+            # there were multiple sensors (with different old tariffs, up to 3),
+            # so we leave just one and remove the others
+            ent_reg: EntityRegistry = await async_get_registry(hass)
+            entity_filter = filter(
```
Please use a generator expression instead of filter + lambda.

MartinHjelmare commented on homeassistant/components/pvpc_hourly_pricing/__init__.py:
```diff
@@ -52,3 +73,54 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+        except ValueError:
+            # there were multiple sensors (with different old tariffs, up to 3),
+            # so we leave just one and remove the others
+            ent_reg: EntityRegistry = await async_get_registry(hass)
```
Please import `async_get` from the entity registry helper and use instead.

MartinHjelmare commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -33,14 +33,29 @@ async def async_setup_entry(
+
+async def options_update_listener(
+    hass: HomeAssistant, entry: config_entries.ConfigEntry
+):
```
Missing return value typing. Please type the whole signature if adding type annotations.

MartinHjelmare commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -33,14 +33,29 @@ async def async_setup_entry(
+        hass.config_entries.async_update_entry(
+            entry, data={**entry.data, **entry.options}
+        )
+        hass.async_create_task(hass.config_entries.async_reload(entry.entry_id))
```
We can await the reload.

MartinHjelmare commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -73,15 +80,24 @@ async def async_added_to_hass(self):
         )
+        # Register callback for when config entry is updated.
+        self.async_on_remove(
+            self._config_entry.add_update_listener(options_update_listener)
```
Registering the entry update listener should be done in `async_setup_entry` in the init module.

MartinHjelmare commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -33,14 +33,29 @@ async def async_setup_entry(
+        power_valley=config_entry.data[ATTR_POWER_P3],
         local_timezone=hass.config.time_zone,
         websession=async_get_clientsession(hass),
         logger=_LOGGER,
```
Side note: The library should not use the Home Assistant logger. The library should use its own logger(s).

azogue replied:
Ok, noted!

I did that a long time ago and I don't remember why 😅. It's an optional parameter, same as the websession, removing it

azogue committed:
Add more typing, don't pass logger obj, move update listener to init

azogue committed:
Adapt migration logic to not increase the config entry version

azogue committed:
Move migration logic to async_setup_entry, avoiding version increase

azogue committed:
Minor change on logging

to get 1 warning for 1st sensor migration, and 1 error for each of the
extra sensors, if there are any, which get completely removed

azogue committed:
Add unit test for migration logic for >1 config entries

MartinHjelmare commented on homeassistant/components/pvpc_hourly_pricing/__init__.py:
```diff
@@ -45,10 +66,67 @@ async def async_setup(hass: HomeAssistant, config: dict):
+                entry.data[ATTR_TARIFF],
+                _DEFAULT_TARIFF,
+            )
+        except ValueError:
```
When does the ValueError happen?

azogue replied:
Hi @MartinHjelmare,

When there is >1 old sensors (very strange, as it has no practical application far from data monitoring, but it can be up to 3), in the conversion to the new tariff only the 1st survives, and for the others, because of the **new unique-id**, the entity_registry fails with that ValueError (printed here):

```
2021-06-15 08:30:13 WARNING (MainThread) [homeassistant.components.pvpc_hourly_pricing] Migrating PVPC sensor from old tariff 'discrimination' to new '2.0TD'. Configure the integration to set your contracted power, and select prices for Ceuta/Melilla, if that is your case
2021-06-15 08:30:13 ERROR (MainThread) [homeassistant.components.pvpc_hourly_pricing] Unique id '2.0TD' is already in use by 'sensor.pvpc'
Traceback (most recent call last):
  File "/Volumes/WORK/contrib/home-assistant/homeassistant/components/pvpc_hourly_pricing/__init__.py", line 86, in async_setup_entry
    await async_migrate_entries(hass, entry.entry_id, update_unique_id)
  File "/Volumes/WORK/contrib/home-assistant/homeassistant/helpers/entity_registry.py", line 744, in async_migrate_entries
    ent_reg.async_update_entity(entry.entity_id, **updates)
  File "/Volumes/WORK/contrib/home-assistant/homeassistant/helpers/entity_registry.py", line 384, in async_update_entity
    return self._async_update_entity(
  File "/Volumes/WORK/contrib/home-assistant/homeassistant/helpers/entity_registry.py", line 457, in _async_update_entity
    raise ValueError(
ValueError: Unique id '2.0TD' is already in use by 'sensor.pvpc'
2021-06-15 08:30:13 ERROR (MainThread) [homeassistant.components.pvpc_hourly_pricing] Old PVPC-2 Sensor sensor.pvpc_extra is removed (another one already exists, using the same tariff)
```

which is the signal to remove the extra config entry with its extra sensor. This was the thing I had problems to reproduce on tests (I forgot to mock the entity-registry), but now it is solved and tested in `test_multi_sensor_migration`

About your comment about the log level for those:
> This is a warning, not an error.

I don't know the best way of doing this: I put an error there because the 'successful migration' message is already a warning ;-), as I wanted to point out to the user that the migrated sensor now can be configured with the options-flow, but I thought about generating a **persistent-notification** instead (is that a good idea?). In that case, success would generate the notification, and for 2nd and 3rd sensor, it would be just a warning (+ removal of the extra integration). I understand that errors should be used _only_ on non-desired/expected behaviours, but I didn't know how to differentiate those 2 possible outcomes (conversion ok vs removal of extra sensors). 

I could also reduce that to warning and finish with it, obviously, so whatever you consider best ;-)

MartinHjelmare replied:
I think both log messages should be warning level. That's fine.

MartinHjelmare replied:
Do we know which of the three existing entity entries should be removed before we start the migration? Then we could just remove the two that we don't want and then migrate the remaining one.

azogue replied:
> I think both log messages should be warning level. That's fine.

Ok, I'll change it

> Do we know which of the three existing entity entries should be removed before we start the migration? Then we could just remove the two that we don't want and then migrate the remaining one.

It doesn't matter. Neither of the 3 old tariffs have continuity into the new one (which is unique, except for a specific region of Spain, where it's another _variant_). That's why I did it so the first one survives and the rest, if any, are removed. And as said before, this is a very strange scenario (a user with >1 sensors), and, anyway, all old sensors working right now are showing now the same price (so they are in fact the same now. This change in the law affects all users of the old 3 tariffs, which now become just one

MartinHjelmare replied:
If this case happens we should remove the surplus entity entries and finish the migration for the remaining entry. Currently, we remove one entry and then return False for config entry setup. That's not the correct solution.

MartinHjelmare commented on homeassistant/components/pvpc_hourly_pricing/__init__.py:
```diff
@@ -45,10 +66,67 @@ async def async_setup(hass: HomeAssistant, config: dict):
+            for entity_id, reg_entry in ent_reg.entities.items():
+                if reg_entry.config_entry_id == entry.entry_id:
+                    ent_reg.async_remove(entity_id)
+                    _LOGGER.error(
```
This is a warning, not an error.

MartinHjelmare commented on tests/components/pvpc_hourly_pricing/test_sensor.py:
```diff
@@ -85,3 +109,65 @@ def mock_now():
+
+            # check migration with removal of extra sensors
+            assert len(entity_reg.entities) == 1
+            assert entity1.entity_id in entity_reg.entities.keys()
```
The keys is the default used for dict membership test.
```suggestion
            assert entity1.entity_id in entity_reg.entities
```

MartinHjelmare commented on tests/components/pvpc_hourly_pricing/test_sensor.py:
```diff
@@ -85,3 +109,65 @@ def mock_now():
+    caplog.clear()
+    with caplog.at_level(logging.WARNING):
+        with patch("homeassistant.util.dt.utcnow", new=mock_now):
+            await hass.config_entries.async_setup(config_entry_1.entry_id)
```
We can assert this.
```suggestion
            assert await hass.config_entries.async_setup(config_entry_1.entry_id)
```

MartinHjelmare commented on tests/components/pvpc_hourly_pricing/test_sensor.py:
```diff
@@ -85,3 +109,65 @@ def mock_now():
+        with patch("homeassistant.util.dt.utcnow", new=mock_now):
+            await hass.config_entries.async_setup(config_entry_1.entry_id)
+            assert len(caplog.messages) == 2
+            assert_setup_component(1, DOMAIN)
```
And remove this.
```suggestion
```

azogue committed:
Minor changes on tests

azogue committed:
Remove outdated YAML example from docstring of deprecated async_setup

azogue committed:
Emit warning instead of error on old sensor discard

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
We should mention the potential breakage for users that have multiple config entries and sensors. Please add a breaking change paragraph in the PR description with info about that change and what those users need to do to cope with the change, ie updating automations and scripts that may use the removed sensors.

azogue commented:
Hi @MartinHjelmare, thank you for reviewing this 🍻

> Please add a breaking change paragraph in the PR description

Is this ok? (feel free to change it, as I'm not the best writer in english 🙈😜)

> With the change to the new, and unique, electric tariff 2.0TD, if you previously had configured multiple PVPC sensors monitoring prices for more than one of the old tariffs, only the first one will survive, so if you have any automations or scripts that depend on these removed sensors, you might need to adjust them.

MartinHjelmare commented:
Great!
